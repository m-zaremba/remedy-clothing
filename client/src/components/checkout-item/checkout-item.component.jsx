import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { clearItemFromCart, addItem, removeItem } from '../../redux/cart/cart.actions';

import {
  CheckoutItemContainer,
  CheckoutImageContainer,
  Name,
  Price,
  Quantity,
  Arrow,
  Value,
  RemoveButton,
} from './checkout-item.styles';

export const CheckoutItem = ({ cartItem, clearItem, addItem, removeItem }) => {
  const { imageUrl, name, quantity, price } = cartItem;
  return (
    <CheckoutItemContainer>
      <CheckoutImageContainer>
        <img src={imageUrl} alt="item" />
      </CheckoutImageContainer>
      <Name>{name}</Name>
      <Quantity>
        <Arrow
          onClick={() => removeItem(cartItem)}
          onKeyDown={() => removeItem(cartItem)}
          role="button"
          tabIndex={0}
        >
          &#10094;
        </Arrow>
        <Value>{quantity}</Value>
        <Arrow
          onClick={() => addItem(cartItem)}
          onKeyDown={() => addItem(cartItem)}
          role="button"
          tabIndex={0}
        >
          &#10095;
        </Arrow>
      </Quantity>
      <Price>{price}</Price>
      <RemoveButton
        onClick={() => clearItem(cartItem)}
        onKeyDown={() => clearItem(cartItem)}
        role="button"
        tabIndex={0}
      >
        &#10005;
      </RemoveButton>
    </CheckoutItemContainer>
  );
};

const mapDispatchToProps = (dispatch) => ({
  clearItem: (item) => dispatch(clearItemFromCart(item)),
  addItem: (item) => dispatch(addItem(item)),
  removeItem: (item) => dispatch(removeItem(item)),
});

CheckoutItem.propTypes = {
  cartItem: PropTypes.shape({
    imageUrl: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    quantity: PropTypes.number,
    price: PropTypes.number,
  }).isRequired,
  clearItem: PropTypes.func.isRequired,
  addItem: PropTypes.func.isRequired,
  removeItem: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(CheckoutItem);
