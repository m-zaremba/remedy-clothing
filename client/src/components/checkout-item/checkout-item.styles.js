import styled, { css } from 'styled-components';

const descriptionStyles = css`
  width: 23%;
`;

export const CheckoutItemContainer = styled.div`
  width: 100%;
  display: flex;
  min-height: 100px;
  border-bottom: 1px solid darkgrey;
  padding: 15px 0;
  font-size: 20px;
  align-items: center;

  @media screen and (max-width: 800px) {
    font-size: 16px;
  }
`;

export const CheckoutImageContainer = styled.div`
  width: 23%;
  padding-right: 15px;

  img {
    width: 100%;
    height: 100%;
  }
`;

export const Name = styled.div`
  ${descriptionStyles}
`;

export const Price = styled.div`
  ${descriptionStyles}
`;

export const Quantity = styled.div`
  ${descriptionStyles}
  display: flex;
`;

export const Arrow = styled.div`
  cursor: pointer;
`;

export const Value = styled.span`
  margin: 0 10px;
`;

export const RemoveButton = styled.div`
  padding-left: 12px;
  cursor: pointer;

  @media screen and (max-width: 800px) {
    padding-left: 0;
  }
`;
