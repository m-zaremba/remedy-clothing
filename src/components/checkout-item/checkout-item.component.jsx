import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { clearItemFromCart, addItem, removeItem } from '../../redux/cart/cart.actions';

import './checkout-item.styles.scss';

const CheckoutItem = ({ cartItem, clearItem, addItem, removeItem }) => {
  const { imageUrl, name, quantity, price } = cartItem;
  return (
    <div className="checkout-item">
      <div className="image-container">
        <img src={imageUrl} alt="item" />
      </div>
      <span className="name">{name}</span>
      <span className="quantity">
        <div
          className="arrow"
          onClick={() => removeItem(cartItem)}
          onKeyDown={() => removeItem(cartItem)}
          role="button"
          tabIndex={0}
        >
          &#10094;
        </div>
        <span className="value">{quantity}</span>
        <div
          className="arrow"
          onClick={() => addItem(cartItem)}
          onKeyDown={() => addItem(cartItem)}
          role="button"
          tabIndex={0}
        >
          &#10095;
        </div>
      </span>
      <span className="price">{price}</span>
      <div
        className="remove-button"
        onClick={() => clearItem(cartItem)}
        onKeyDown={() => clearItem(cartItem)}
        role="button"
        tabIndex={0}
      >
        &#10005;
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => ({
  clearItem: (item) => dispatch(clearItemFromCart(item)),
  addItem: (item) => dispatch(addItem(item)),
  removeItem: (item) => dispatch(removeItem(item)),
});

CheckoutItem.propTypes = {
  cartItem: PropTypes.shape({
    imageUrl: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    quantity: PropTypes.number,
    price: PropTypes.number,
  }).isRequired,
  clearItem: PropTypes.func.isRequired,
  addItem: PropTypes.func.isRequired,
  removeItem: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(CheckoutItem);
