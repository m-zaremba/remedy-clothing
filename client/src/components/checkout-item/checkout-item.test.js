import React from 'react';
import { shallow } from 'enzyme';

import { CheckoutItem } from './checkout-item.component';

describe('testing CheckoutItem component', () => {
  let wrapper;
  let mockClearItem;
  let mockAddItem;
  let mockRemoveItem;

  beforeEach(() => {
    mockClearItem = jest.fn();
    mockAddItem = jest.fn();
    mockRemoveItem = jest.fn();

    const mockProps = {
      cartItem: {
        imageUrl: 'www.testURL.com',
        price: 10,
        name: 'testItem',
        quantity: 1,
      },
      clearItem: mockClearItem,
      addItem: mockAddItem,
      removeItem: mockRemoveItem,
    };

    wrapper = shallow(<CheckoutItem {...mockProps} />);
  });

  it('should render CheckoutItem component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call clearItem when remove button is clicked', () => {
    wrapper.find('RemoveButton').simulate('click');
    expect(mockClearItem).toHaveBeenCalled();
  });

  it('should call removeItem when left arrow is clicked', () => {
    wrapper.find('Quantity').childAt(0).simulate('click');

    expect(mockRemoveItem).toHaveBeenCalled();
  });

  it('should call addItem when right arrow is clicked', () => {
    wrapper.find('Quantity').childAt(2).simulate('click');

    expect(mockAddItem).toHaveBeenCalled();
  });
});
