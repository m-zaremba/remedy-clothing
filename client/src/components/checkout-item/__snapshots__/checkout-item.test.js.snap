// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CheckoutItem component should render CheckoutItem component 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="item"
      src="www.testURL.com"
    />
  </styled.div>
  <styled.div>
    testItem
  </styled.div>
  <Quantity>
    <styled.div
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      ❮
    </styled.div>
    <styled.span>
      1
    </styled.span>
    <styled.div
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      ❯
    </styled.div>
  </Quantity>
  <styled.div>
    10
  </styled.div>
  <RemoveButton
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    ✕
  </RemoveButton>
</styled.div>
`;
